File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-6.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} list1
     * @param {ListNode} list2
     * @return {ListNode}
     */

    mergeTwoListsRec(temp, list1, list2) {
        if(list1 == null && list2 == null){
            return;
        }
        else if(list1 != null && list2 == null){
            temp.next = list1;
            return;
        }
        else if(list2 != null && list1 == null){
            temp.next = list2;
            return;
        }

        if(list1.val < list2.val){
           temp.next = list1;
           list1 = list1.next;
        }
        else{
           temp.next = list2;
           list2 = list2.next;
        }

        return this.mergeTwoListsRec(temp.next, list1, list2)
    }

    mergeTwoLists(list1, list2){
        const dummy = new ListNode();
        let temp  = dummy;
        this.mergeTwoListsRec(temp, list1, list2);
        return dummy.next;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-6.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    #Q: is n gonna be valid? assum yes
    #P: add a dumy node and slide window to find n+1 node from end (prev)
    #P: then you can do prev.next = prev.next.next\
    # D->1->2
    #    L  R
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        #insert dummy node as head to find n+1 node
        dummy = ListNode()
        dummy.next = head

        #create a window of size n
        L, R = dummy, dummy
        for _ in range(n):
            R = R.next
        
        #slide window until R is at last node (L will be n+1th node form end)
        while R.next:
            R = R.next
            L = L.next
        
        #delete the nth node
        L.next = L.next.next

        #return new head
        return dummy.next